crates/subspace-networking/src/peer_info/handler.rs

```diff
@@ -0,0 +1,281 @@
+use libp2p::core::UpgradeError;
+use libp2p::swarm::handler::{
+    ConnectionEvent, DialUpgradeError, FullyNegotiatedInbound, FullyNegotiatedOutbound,
+    ListenUpgradeError,
+};
+use libp2p::swarm::{
+    ConnectionHandler, ConnectionHandlerEvent, ConnectionHandlerUpgrErr, KeepAlive,
+    NegotiatedSubstream, SubstreamProtocol,
+};
+use std::error::Error;
+use std::task::{Context, Poll};
+use std::time::Duration;
+use std::{fmt, io};
+use tracing::debug;
+
+/// The configuration for peer-info protocol.
+#[derive(Debug, Clone)]
+pub struct Config {
+    /// Protocol timeout.
+    timeout: Duration,
+
+    /// Protocol name.
+    protocol_name: &'static [u8],
+}
+
+impl Config {
+    /// Creates a new [`Config`] with the following default settings:
+    ///
+    ///   * [`Config::with_timeout`] 20s
+    pub fn new(protocol_name: &'static [u8]) -> Self {
+        Self {
+            timeout: Duration::from_secs(20),
+            protocol_name,
+        }
+    }
+
+    /// Sets the protocol timeout.
+    pub fn with_timeout(mut self, d: Duration) -> Self {
+        self.timeout = d;
+        self
+    }
+}
+
+/// The successful result of processing an inbound or outbound peer info requests.
+#[derive(Debug)]
+pub enum PeerInfoSuccess {
+    Received(PeerInfo),
+    DataSent,
+}
+
+/// A peer info protocol failure.
+#[derive(Debug)]
+pub enum PeerInfoError {
+    /// The peer does not support the peer info protocol.
+    Unsupported,
+    /// The peer info request failed.
+    Other {
+        error: Box<dyn Error + Send + 'static>,
+    },
+}
+
+impl fmt::Display for PeerInfoError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            PeerInfoError::Other { error } => write!(f, "Peer info error: {error}"),
+            PeerInfoError::Unsupported => write!(f, "Peer info protocol not supported"),
+        }
+    }
+}
+
+impl Error for PeerInfoError {
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        match self {
+            PeerInfoError::Other { error } => Some(&**error),
+            PeerInfoError::Unsupported => None,
+        }
+    }
+}
+
+/// Protocol handler that handles peer-info requests.
+///
+/// Any protocol failure produces an error that closes the connection.
+pub struct Handler {
+    /// Configuration options.
+    config: Config,
+    /// The outbound request state.
+    outbound: Option<OutboundState>,
+    /// The inbound request future.
+    inbound: Option<InPeerInfoFuture>,
+    /// Last peer-info error.
+    error: Option<PeerInfoError>,
+}
+
+impl Handler {
+    /// Builds a new [`Handler`] with the given configuration.
+    pub fn new(config: Config) -> Self {
+        Handler {
+            config,
+            outbound: None,
+            inbound: None,
+            error: None,
+        }
+    }
+}
+
+/// Marker struct for outbound peer-info requests.
+#[derive(Debug, Clone)]
+pub struct HandlerInEvent {
+    pub peer_info: PeerInfo,
+}
```

Comment:
Try not to put structs in between definition and impls of other structs (`Handler` in this case)